go/pkg/client/cas.go

```diff
@@ -1106,56 +1135,69 @@ func (c *Client) FlattenActionOutputs(ctx context.Context, ar *repb.ActionResult
 }
 
 // DownloadDirectory downloads the entire directory of given digest.
-func (c *Client) DownloadDirectory(ctx context.Context, d digest.Digest, execRoot string, cache filemetadata.Cache) (map[string]*TreeOutput, error) {
+// It returns the number of logical and real bytes downloaded, which may be different from sum
+// of sizes of the files due to dedupping and compression.
+func (c *Client) DownloadDirectory(ctx context.Context, d digest.Digest, execRoot string, cache filemetadata.Cache) (map[string]*TreeOutput, MovedBytesMetadata, error) {
```

Comment:
Return a *MovedBytesMetadata here and elsewhere. That way, you initialize them with &MovedBytesMetadata{...}, too. Basically, you only ever deal with pointers.